cloudpickle/cloudpickle_fast.py

```diff
@@ -0,0 +1,541 @@
+
+    Set protocol=pickle.DEFAULT_PROTOCOL instead if you need to ensure
+    compatibility with older versions of Python.
+    """
+    with io.BytesIO() as file:
+        cp = CloudPickler(file, protocol=protocol)
+        cp.dump(obj)
+        return file.getvalue()
+
+
+# COLLECTION OF OBJECTS __getnewargs__-LIKE METHODS
+# -------------------------------------------------
+
+def _class_getnewargs(obj):
+    # On PyPy, __doc__ is a readonly attribute, so we need to include it in
+    # the initial skeleton class.  This is safe because we know that the
+    # doc can't participate in a cycle with the original class.
+    type_kwargs = {'__doc__': obj.__dict__.get('__doc__', None)}
+
+    if hasattr(obj, "__slots__"):
+        type_kwargs["__slots__"] = obj.__slots__
+
+    __dict__ = obj.__dict__.get('__dict__', None)
+    if isinstance(__dict__, property):
+        type_kwargs['__dict__'] = __dict__
+
+    return (type(obj), obj.__name__, obj.__bases__, type_kwargs,
+            _ensure_tracking(obj), None)
+
+
+def _enum_getnewargs(obj):
+    members = dict((e.name, e.value) for e in obj)
+    return (obj.__bases__, obj.__name__, obj.__qualname__, members,
+            obj.__module__, _ensure_tracking(obj), None)
+
+
+# COLLECTION OF OBJECTS RECONSTRUCTORS
+# ------------------------------------
+def _file_reconstructor(retval):
+    return retval
+
+
+# COLLECTION OF OBJECTS STATE GETTERS
+# -----------------------------------
+def _function_getstate(func):
+    # - Put func's dynamic attributes (stored in func.__dict__) in state. These
+    #   attributes will be restored at unpickling time using
+    #   f.__dict__.update(state)
+    # - Put func's members into slotstate. Such attributes will be restored at
+    #   unpickling time by iterating over slotstate and calling setattr(func,
+    #   slotname, slotvalue)
+    slotstate = {
+        "__name__": func.__name__,
+        "__qualname__": func.__qualname__,
+        "__annotations__": func.__annotations__,
+        "__kwdefaults__": func.__kwdefaults__,
+        "__defaults__": func.__defaults__,
+        "__module__": func.__module__,
+        "__doc__": func.__doc__,
+        "__closure__": func.__closure__,
+    }
+
+    f_globals_ref = _extract_code_globals(func.__code__)
+    f_globals = {k: func.__globals__[k] for k in f_globals_ref if k in
+                 func.__globals__}
+
+    closure_values = (
+        list(map(_get_cell_contents, func.__closure__))
+        if func.__closure__ is not None else ()
+    )
+
+    # Extract submodules referenced by attribute lookup (no global opcode).
+    # Storing the loaded submodules in a smoke __submodule__ attribute of
+    # func.__globals__ allow these modules to be saved at pickling time, and
+    # thus imported when the function is unpickled.
+    f_globals["__submodules__"] = _find_loaded_submodules(
+        func.__code__, itertools.chain(f_globals.values(), closure_values))
```

Comment:
Also maybe the name "subimports" is actually more descriptive than "submodules". The point is that we want to trigger the side effect of importing those nested references to modules/packages. WDYT of the following names:

- `_find_loaded_submodules` => `_find_dependent_subimports`
- `f_globals["__submodules__"]` => `f_globals["_cloudpickle_subimports"]`

Reply:
I like  `_find_loaded_submodules`/`_find_imported_submodules` because this is what this function does.  Explaining the very important point you are stating ("trigger the side effect of importing those nested references to modules/packages") in a comment is descriptive enough to me. I still changed the code according to your suggestion. Feel free to resolve the convo if you don't change your mind.

Reply:
I like `_find_imported_submodules` better then and then `f_globals["_cloudpickle_submodules"]`.